src/NuGet.Core/NuGet.DependencyResolver.Core/ResolverUtility.cs

```diff
@@ -130,7 +130,7 @@ public static class ResolverUtility
                 Data = new RemoteResolveResult
                 {
                     Match = match,
-                    Dependencies = dependencies.Dependencies
+                    Dependencies = dependencies.Dependencies.ToList()
```

Comment:
I don't know if I am missing any internals here, AFAIK `.ToList()` is an `O(n)` operation.  Looking for the references of `LibraryDependencyInfo.Dependencies`, I see from the following implementation that `.ToArray()` has been invoked but in all the other places collection is maintained in a `List` object.

https://github.com/NuGet/NuGet.Client/blob/7721feed7c830d5d11f6c43a2205888f36d7e4ab/src/NuGet.Core/NuGet.Commands/RestoreCommand/SourceRepositoryDependencyProvider.cs#L475

My proposal is as follows.

-  Change the above code to invoke `.ToList()` instead of `.ToArray()`
-  Let `RemoteResolveResult.Dependencies` be `IEnumerable<LibraryDependency>`
- Revert other changes because of the following implementation
https://github.com/NuGet/NuGet.Client/blob/7721feed7c830d5d11f6c43a2205888f36d7e4ab/build/Shared/SharedExtensions.cs#L17-L26

Reply:
@kartheekp-ms 

We're trying to avoid unintentional boxing. 
If we call AsList every time we enumerate, we'd fix this occurrence, but doesn't mean we won't run into the same issue again.

Reply:
I agree but we are invoking `.ToList()` method in few places so that the dependencies are maintained as a `List`. IMO, calling `.AsList()` provides a better performance if the collection is a list. Now that we know `List.GetEnumerator()` avoids unintentional boxing, can we follow below steps to avoid running into the same issue again.

- Add about benefit of using `List.GetEnumerator()` method in our [coding guidelines doc](https://github.com/NuGet/NuGet.Client/blob/dev/docs/coding-guidelines.md).
- Check the memory allocations for hot code paths on a periodic basis.

Reply:
Calling AsList instead of ToList is a good change where appropriate. 

We can consider adding it to the coding guidelines, but the specificity of things make me think it won't solve everything. 

We should be looking at the allocations in our perf tests.

The beauty of these changes is that JP is making that once discovered, they are relatively low effort. I think we should keep that spirit for future changes. :)